stdlib/pqueue.ml

```diff
@@ -0,0 +1,223 @@
+    let create =
+      Dynarray.create
+
+    let length =
+      Dynarray.length
+
+    let is_empty h =
+      length h = 0
+
+    let clear =
+      Dynarray.clear
+
+    (* The node at index [i] has children nodes at indices [2 * i + 1]
+       and [2 * i + 2] -- if they are valid indices in the dynarray. *)
+    let left_child i = 2 * i + 1
+    let right_child i = 2 * i + 2
+    let parent_node i = (i - 1) / 2
+
+    (* We say that a heap respects the "heap ordering" if the value of
+       each node is no greater than the value of its children. The
+       algorithm manipulates arrays that respect the heap algorithm,
+       except for one node whose value may be too small or too large.
+
+       The auxiliary functions [sift_up] and [sift_down] take
+       such a misplaced value, and move it "up" (respectively: "down")
+       until the heap ordering is restored.
+
+       Functions [sift_up] and [sift_down] do not perform swaps, but
+       rather expect the value to be assigned in the heap as an
+       additional parameter [x], resulting in twice less assignments. *)
+
+    (* store [x] at index [i], moving it up if necessary *)
+    let rec sift_up h i x =
+      if i = 0 then Dynarray.set h 0 x else
+      let p = parent_node i in
+      let y = Dynarray.get h p in
+      if E.compare x y < 0 then (
+        Dynarray.set h i y;
+        sift_up h p x
+      ) else
+        Dynarray.set h i x
+
+    let add h x =
+      let i = Dynarray.length h in
+      Dynarray.add_last h x;
+      if i > 0 then sift_up h i x
+
+    let add_seq h s =
+      Seq.iter (add h) s
+
+    exception Empty
+
+    let min_elt h =
+      if Dynarray.length h = 0 then raise Empty;
+      Dynarray.get h 0
+
+    let min_elt_opt h =
+      if Dynarray.length h = 0 then None else Some (Dynarray.get h 0)
+
+    let lt h i j =
+      E.compare (Dynarray.get h i) (Dynarray.get h j) < 0
+
+    (* store [x] at index [i], moving it down if necessary *)
+    let rec sift_down h ~len i x =
+      let left = left_child i in
+      if left >= len then Dynarray.set h i x (* no child, stop *) else
+      let smallest =
+        let right = right_child i in
+        if right >= len then left (* no right child *) else
+        if lt h left right then left else right
+      in
+      let y = Dynarray.get h smallest in
+      if E.compare y x < 0 then (
+        Dynarray.set h i y;
+        sift_down h ~len smallest x
+      ) else
+        Dynarray.set h i x
+
+    let pop_min h =
+      let n = Dynarray.length h in
+      if n = 0 then raise Empty;
+      let x = Dynarray.pop_last h in
+      if n = 1 then x else (
+        let r = Dynarray.get h 0 in
+        sift_down h ~len:(n - 1) 0 x;
+        r
+      )
+
+    let pop_min_opt h =
+      let n = Dynarray.length h in
+      if n = 0 then None else
+      let x = Dynarray.pop_last h in
+      if n = 1 then Some x else (
+        let r = Dynarray.get h 0 in
+        sift_down h ~len:(n - 1) 0 x;
+        Some r
+      )
```

Comment:
I am not sure if this is worth duplicating the code instead of using `try Some (pop_min h) with Empty -> None`. Both are used in the stdlib.

Reply:
(Surely you mean `match pop_min h with x -> Some x | exception Empty -> None`.)

Reply:
In general it is a nice property of code that does not raise exceptions as part of its interface to also not clobber the existing exception backtrace. The code as written has this property, but since `pop_min` does not, and should not, use `raise_notrace`, then defining the option variant from the exception variant would sacrifice this property.